pallets/phala-world/src/nft_sale.rs

```diff
@@ -0,0 +1,1722 @@
+					ensure!(id != PreorderId::max_value(), Error::<T>::NoAvailablePreorderId);
+					*n += 1;
+					Ok(id)
+				})?;
+			// Verify metadata
+			let metadata = Self::get_empty_metadata();
+			let preorder = PreorderInfo {
+				owner: sender.clone(),
+				race,
+				career,
+				metadata,
+				preorder_status: PreorderStatus::Pending,
+			};
+			// Reserve currency for the preorder at the Hero origin_of_shell price
+			<T as pallet::Config>::Currency::reserve(&sender, T::HeroOriginOfShellPrice::get())?;
+
+			Preorders::<T>::insert(preorder_id, preorder);
+
+			Self::deposit_event(Event::OriginOfShellPreordered { owner: sender, preorder_id });
+
+			Ok(())
+		}
+
+		/// This is an admin only function that will be used to set a preorder's status from Pending
+		/// to either Chosen or NotChosen
+		///
+		/// Parameters:
+		/// `origin`: Expected to come from Overlord admin account
+		/// `preorder_id`: Preorder id to change preorder status
+		/// `status`: Either Chosen or NotChosen
+		#[pallet::weight(10_000 + T::DbWeight::get().reads_writes(1,1))]
+		#[transactional]
+		pub fn set_preorder_status(
+			origin: OriginFor<T>,
+			preorder_statuses: Vec<(PreorderId, PreorderStatus)>,
+		) -> DispatchResult {
+			// Ensure Overlord account makes call
+			let sender = ensure_signed(origin)?;
+			Self::ensure_overlord(sender)?;
+			// Get iter limit
+			let iter_limit = T::IterLimit::get();
+			let mut index = 0;
+			// Iterate through the preorder_statuses Vec
+			for (preorder_id, preorder_status) in preorder_statuses {
+				if index < iter_limit {
+					// Ensure value exists
+					ensure!(
+						Preorders::<T>::contains_key(preorder_id),
+						Error::<T>::NoAvailablePreorderId
+					);
+					// Change status of preorder and add to PreorderResult
+					let mut preorder_info = Preorders::<T>::take(preorder_id)
+						.ok_or(Error::<T>::NoAvailablePreorderId)?;
+					preorder_info.preorder_status = preorder_status;
+					match preorder_status {
+						PreorderStatus::Chosen => ChosenPreorders::<T>::insert(
+							&preorder_info.owner,
+							preorder_id,
+							&preorder_info,
+						),
+						PreorderStatus::NotChosen => NotChosenPreorders::<T>::insert(
+							&preorder_info.owner,
+							preorder_id,
+							&preorder_info,
+						),
+						PreorderStatus::Pending => return Err(Error::<T>::PreorderIsPending.into()),
+					}
+
+					Self::deposit_event(Event::PreorderResultChanged {
+						preorder_id,
+						status: preorder_status,
+					});
+					index += 1;
+				} else {
+					// Break from iterator to ensure block size doesn't grow too large. Re-call this
+					// function and it will start from where it left off.
+					break
+				}
+			}
+
+			Ok(())
+		}
+
+		/// Claim function for bulk minting chosen preorders for the sender
+		///
+		/// Parameters:
+		/// `origin`: Sender
+		#[pallet::weight(10_000 + T::DbWeight::get().reads_writes(1,1))]
+		#[transactional]
+		pub fn claim_chosen_preorders(origin: OriginFor<T>) -> DispatchResult {
```

Comment:
I suggest to allow anyone to call this function for someone (themself or someone else)

Reply:
Why would we allow someone to claim a preorder for someone else? Based on the logic, the NFT would be minted to whoever calls this function.

Reply:
Then we can trigger the claim for someone else? I will ask the PM if this is needed